packages/react-jss/src/JssProvider.js

```diff
@@ -20,85 +20,88 @@ type Props = {
   }
 
-  static childContextTypes = contextTypes
+  managers: {} = {}
 
-  static contextTypes = contextTypes
+  generateId = createGenerateId()
 
-  // JssProvider can be nested. We allow to overwrite any context prop at any level.
-  // 1. Check if there is a value passed over props.
-  // 2. If value was passed, we set it on the child context.
-  // 3. If value was not passed, we proxy parent context (default context behaviour).
-  getChildContext(): Context {
+  createContext(outerContext: Context): Context {
     const {
       registry,
       classNamePrefix,
-      jss: localJss,
+      jss,
       generateId,
-      disableStylesGeneration
+      disableStylesGeneration,
+      sheetOptions
     } = this.props
-    // eslint-disable-next-line react/react/destructuring-assignment
-    const sheetOptions = this.context[ns.sheetOptions] || {}
-    const context: Context = {[ns.sheetOptions]: sheetOptions}
+    // Clone the outer context
+    const context = {
+      ...outerContext,
+      managers: this.managers
+    }
 
-    if (registry) {
-      context[ns.sheetsRegistry] = registry
-      // This way we identify a new request on the server, because user will create
-      // a new Registry instance for each.
-      if (registry !== this.registry) {
-        // We reset managers because we have to regenerate all sheets for the new request.
-        this.managers = {}
-        this.registry = registry
+    if (sheetOptions) {
+      warning(
+        'generateId' in sheetOptions,
+        '[JssProvider] Do not pass the generateId inside the sheet options! We have a special prop for that!'
+      )
+
+      context.sheetOptions = {
+        ...context.sheetOptions,
+        ...sheetOptions
       }
     }
 
-    // Make sure we don't loose managers when JssProvider is used inside of a stateful
-    // component which decides to rerender.
-    context[ns.managers] = this.managers
+    if (registry) {
+      context.registry = registry
+    }
 
     if (generateId) {
-      sheetOptions.generateId = generateId
-    } else if (!sheetOptions.generateId) {
-      if (!this.generateId) {
-        let createGenerateId = createGenerateIdDefault
-        if (localJss && localJss.options.createGenerateId) {
-          createGenerateId = localJss.options.createGenerateId
-        }
-        // Make sure we don't loose the generator when JssProvider is used inside of a stateful
-        // component which decides to rerender.
-        this.generateId = createGenerateId()
-      }
+      context.sheetOptions.generateId = generateId
+    } else if (!context.sheetOptions.generateId) {
+      context.sheetOptions.generateId = this.generateId
+    }
+
+    // Merge the classname prefix
+    if (classNamePrefix) {
+      context.sheetOptions.classNamePrefix =
+        classNamePrefix + (context.sheetOptions.classNamePrefix || '')
+    }
 
-      sheetOptions.generateId = this.generateId
+    if (jss) {
+      context.jss = jss
```

Comment:
what about jss on outerContextt?

Reply:
It should be possible to overwrite jss correct?

Reply:
I thought it was possible before? At least it would make sense to me right now

Reply:
Now I'm confused. It should be possible to overwrite the jss instance, correct?

Reply:
I get it now, we override in the inner provider the instance on the context if one was passed higher up. Its the same beehavior as it was before